enterprise/server/remote_execution/containers/podman/podman.go

```diff
@@ -443,13 +457,87 @@ func (c *podmanCommandContainer) PullImage(ctx context.Context, creds container.
 
 	ps.mu.Lock()
 	defer ps.mu.Unlock()
+	if *imageStreamingEnabled {
+		if err := c.prepareToStreamImage(ctx); err != nil {
+			return err
+		}
+	}
 	if err := c.pullImage(ctx, creds); err != nil {
 		return err
 	}
 	ps.pulled = true
 	return nil
 }
 
+func (c *podmanCommandContainer) prepareToStreamImage(ctx context.Context) error {
+	ctx, err := prefix.AttachUserPrefixToContext(ctx, c.env)
+	if err != nil {
+		return err
+	}
+	resp, err := c.env.GetSociArtifactStoreClient().GetArtifacts(ctx, &socipb.GetArtifactsRequest{Image: c.image})
+	if err != nil {
+		return err
+	}
+	blobReq := repb.BatchReadBlobsRequest{
+		InstanceName:   "soci",
+		DigestFunction: repb.DigestFunction_SHA256,
+	}
+	for _, artifact := range resp.Artifacts {
+		if artifact.Type == socipb.Type_UNKNOWN_TYPE {
+			return status.InternalErrorf("SociArtifactStore returned unknown artifact with hash %s" + artifact.Digest.Hash)
+		} else if artifact.Type == socipb.Type_SOCI_INDEX {
+			// Write the index file, this is what the snapshotter uses to
+			// associate the requested image with its soci index.
+			if err = os.MkdirAll(sociIndexDirectory, 0644); err != nil {
+				return err
+			}
+			sociIndexDigest := godigest.NewDigestFromEncoded(godigest.SHA256, artifact.Digest.Hash)
+			imageDigest := godigest.NewDigestFromEncoded(godigest.SHA256, resp.ImageId)
+			if err = os.WriteFile(sociIndex(imageDigest), []byte(sociIndexDigest.String()), 0644); err != nil {
+				return err
+			}
+		}
+		blobReq.Digests = append(blobReq.Digests, artifact.Digest)
+	}
+	blobResp, err := c.env.GetContentAddressableStorageClient().BatchReadBlobs(ctx, &blobReq)
```

Comment:
IIRC you said some of these artifacts can be large (hundreds of megabytes?)
BatchReadBlobs is not appropriate to use with large artifacts, they need to be retrieved via the streaming API

Reply:
They're probably more like tens of megabytes but I haven't tried a really large image layer yet, so good idea to switch to streaming them.